source/extensions/http/header_validators/envoy_default/path_normalizer.cc

```diff
@@ -0,0 +1,403 @@
+  // be safely decoded, as otherwise the data may be mistaken for component
+  // delimiters. The only exception is for percent-encoded octets corresponding
+  // to characters in the unreserved set, which can be decoded at any time.
+  // SPELLCHECKER(on)
+
+  if (iter == end || *iter != '%') {
+    return {PercentDecodeResult::Invalid};
+  }
+
+  char ch = '\0';
+  // Normalize and decode the octet
+  for (int i = 0; i < 2; ++i) {
+    ++iter;
+    if (iter == end) {
+      return {PercentDecodeResult::Invalid};
+    }
+
+    char nibble = *iter;
+    if (!isxdigit(*iter)) {
+      return {PercentDecodeResult::Invalid};
+    }
+
+    // normalize
+    nibble = nibble >= 'a' ? nibble ^ 0x20 : nibble;
+    *iter = nibble;
+
+    // decode
+    int factor = i == 0 ? 16 : 1;
+    ch += factor * (nibble >= 'A' ? (nibble - 'A' + 10) : (nibble - '0'));
+  }
+
+  if (testChar(kUnreservedCharTable, ch)) {
+    // Based on RFC, only decode characters in the UNRESERVED set.
+    return {PercentDecodeResult::Decoded, ch};
+  }
+
+  if (ch == '/') {
+    // We decoded a slash character and how we handle it depends on the active configuration.
+    switch (config_.uri_path_normalization_options().path_with_escaped_slashes_action()) {
+    case HeaderValidatorConfig_UriPathNormalizationOptions::IMPLEMENTATION_SPECIFIC_DEFAULT:
+      ABSL_FALLTHROUGH_INTENDED;
+    case HeaderValidatorConfig_UriPathNormalizationOptions::KEEP_UNCHANGED:
+      // default implementation: normalize the encoded octet and accept the path
+      return {PercentDecodeResult::Normalized};
+
+    case HeaderValidatorConfig_UriPathNormalizationOptions::REJECT_REQUEST:
+      // Reject the entire request
+      return {PercentDecodeResult::Reject};
+
+    case HeaderValidatorConfig_UriPathNormalizationOptions::UNESCAPE_AND_FORWARD:
+      // Decode the slash and accept the path.
+      return {PercentDecodeResult::Decoded, ch};
+
+    case HeaderValidatorConfig_UriPathNormalizationOptions::UNESCAPE_AND_REDIRECT:
+      // Decode the slash and response with a redirect to the normalized path.
+      return {PercentDecodeResult::DecodedRedirect, ch};
+
+    default:
+      // This should never occur but it's here to make the compiler happy because of the extra
+      // values added by protobuf.
+      break;
+    }
+  }
+
+  // The octet is a valid encoding but it wasn't be decoded because it was outside the UNRESERVED
+  // character set.
+  return {PercentDecodeResult::Normalized};
+}
+
+/*
+ * Find the start index of the previous segment within the path. The previous segment starts at the
+ * first non-slash character after the preceding slash. For example:
+ *
+ *   path = "/hello/world/..";
+ *                  ^    ^-- iterator
+ *                  |-- start of previous segment
+ *
+ * The ``begin`` iterator is returned on error.
+ */
+std::string::iterator findStartOfPreviousSegment(std::string::iterator iter,
```

Comment:
Does this function intentionally handle `/abc///def/..` to return the iterator to `a`?
If so I think the comment should clarify this use case.

It is unclear where begin should point to when this function is called.

Reply:
I've updated the comment to hopefully clarify the function's usage. In your example, the return result would be:

```cpp
path = "/abc///def/..";
//      ^      ^  ^-- current argument
//      |      |-- return value
//      |-- begin
```